ext/oj/dump.c

```diff
@@ -786,12 +799,22 @@ void oj_dump_cstr(const char *str, size_t cnt, bool is_sym, bool escape1, Out ou
         } else {
             size = (size_t)sz;
         }
+        do_unicode_validation = true;
         break;
     }
-    case RailsEsc:
+    case RailsEsc: {
+        long sz;
         cmap = rails_friendly_chars;
-        size = rails_friendly_size((uint8_t *)str, cnt);
+        sz   = rails_friendly_size((uint8_t *)str, cnt);
+        if (sz < 0) {
+            has_hi = true;
+            size   = (size_t)-sz;
+        } else {
+            size = (size_t)sz;
+        }
```

Comment:
Under what circumstances could `sz` ever be less than 0?

Reply:
I suspect your fork might be behind. That's probably the issue. Can you verify?

Reply:
Ah, what I've done is:

* Copy the implementation of `rails_xss_friendly_size` into `rails_friendly_size`, except change the `rails_xss_friendly_chars` lookup array to the `rails_friendly_chars` lookup array
* Dup the implementation of  `case RailsXEsc` into `case RailsEsc`
* `rails_xss_friendly_size` and `rails_friendly_size` return the number of bytes required for the string if there are no chars with code point > 127, or _negative_ of that number if there are such chars; the sign bit is used to sneak an extra bit of information out of the function.

There's a lot of code in common between them, but I didn't feel I understood the essence of this well enough to implement a better factoring of it. I'm hoping you have some suggestions :)

My best idea is...

* Rename `rails_xss_friendly_size` to `size_t required_buffer_size_for_escaped_string(const uint8_t *str, size_t len, const char *cmap, bool *has_hi_out)`. Make it accept the cmap to use as a parameter, essentially, and pass `hi` as an explicit out-param rather than smuggling it out via the sign bit
* Call `required_buffer_size_for_escaped_string` in `case RailsXEsc` and `case RailsEsc` with different cmaps (`rails_xss_friendly_chars` vs `rails_friendly_chars`)

Does this sound reasonable to you?

Reply:
It does.